### PR TITLE
Netpol: Test that outside world Egress can be effectively allowed while blocking all pod ingress

### LINKED ISSUE #96341: Network policies are blocking allowed traffic
**What happened**:
I have set up a network policy that's blocking traffic that it shouldn't
**What you expected to happen**:
Network policies are supposed to block everything that is not allowed. In my case I am allowing egress traffic to everywhere, except the cidr block `192.168.0.0/16`. I can confirm that all traffic to that block is indeed restricted, and mostly anything else is allowed, except for some specific ips (might be affecting the entire `10.0.0.0/8` block as well even though it's not declared as an "except" block, hence should be allowed).

**How to reproduce it (as minimally and precisely as possible)**:
1. install eksctl to launch a cluster in EKS (this is where I am testing) (https://docs.aws.amazon.com/eks/latest/userguide/eksctl.html)
2. deploy a simple cluster:
```
# test-eksctl.yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: basic-cluster
  region: us-west-2

nodeGroups:
  - name: ng-1
    instanceType: m5.large
    desiredCapacity: 3
    volumeSize: 80
```
`eksctl create cluster -f test-eksctl.yaml`

3. install calico network plugin:
`kubectl apply -f https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/v1.7.5/config/v1.7/calico.yaml`

4. create a new namespace:
`kubectl create ns netpol-test`

5.  deploy a test pod in this namespace:
```
# netutils.yaml
apiVersion: v1
kind: Pod
metadata:
  name: netutils-sd1
spec:
  containers:
    - name: my-pod
      image: amouat/network-utils
      command:
      - /bin/sh
      - -c
      - |
        tail -f /dev/null
```
then test basic networking
```
# Deploy the pod
kubectl -n netpol-test apply -f netutils.yaml
 # (This is to grab the cluster ip for the internal kubernetes service, in my case is `10.100.0.1`) 
kubectl get svc -A
# Get a shell in the pod
kubectl -n netpol-test  exec -it  netutils-sd1 bash
# This commands run inside the container
curl -k -I https://10.100.0.1:443
# Which should return 403 forbidden. This is expected, and it purpose is to demonstrate that we can at least hit this ip (we don't care about authentication here, only traffic)
```
6. deploy the following network policy to that namespace:

```
# test-netpol.yaml
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: test-netpol
spec:
  egress:
  - to:
    # This should allow communication between pods within this namespace
    - podSelector: {}
    - ipBlock:
        # We need this because we need pods to have connectivity to outside world, but not to pods living in the same cluster in other namespaces
        cidr: 0.0.0.0/0
       # Because all my pods for other namespace will live in the 192.169.0.0/16 block, 
       # I don't want pods from this namespace to reach other namespaces, 
       # this is what I found the easiest way to block connectivity across namespaces.
        except:
        - 192.168.0.0/16
  # This is to allow privileged namespaces like prometheus/kube-system to reach pods in this namespace.
  ingress:
  - from:
    - ipBlock:
        cidr: 0.0.0.0/0
  podSelector: {}
  policyTypes:
  - Ingress
  - Egress
```
`kubectl -n netpol-test apply -f test-netpol.yaml `
Try to hit the ip again:
```
kubectl -n netpol-test  exec -it  netutils-sd1 bash
# This commands run inside the container
curl -k -I https://10.100.0.1:443
# This till it will hang until it times out. This is not expected, as we are not blocking this private ip space anywhere.
```

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-11T18:14:22Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.9-eks-4c6976", GitCommit:"4c6976793196d70bc5cd29d56ce5440c9473648e", GitTreeState:"clean", BuildDate:"2020-07-17T18:46:04Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: EKS 1.17
- OS (e.g: `cat /etc/os-release`): macos
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
* aws-cni
* calico network plugin
- Others:

### PR DESCRIPTION
**Pull request by jayunit100**
Follow on to https://github.com/kubernetes/kubernetes/pull/91592/ 

**What type of PR is this?**
#96341 Brings up an interesting case wherein we block traffic after allowing everything on 0.0.0.0/0.  Although this is kinda weird as an overall policy, I think its technically valid, and seems to work... although its a little confusing when you read such a policy (normally you'd disallow traffic from the outside and allow on the inside).  I think its a new validation so adding as kind feature.

/kind test

**What this PR does / why we need it**:

Show a clear example of how you can deny internal traffic while allowing external egress traffic in the netpol suite.

**Which issue(s) this PR fixes**: 
Fixes #96341

**Example output**

```

Nov 15 10:29:28.203: INFO: expected:

-               netpol-5383-x/a netpol-5383-x/b netpol-5383-x/c netpol-5383-y/a netpol-5383-y/b netpol-5383-y/c netpol-5383-z/a netpol-5383-z/b netpol-5383-z/c
netpol-5383-x/a X               X               X               X               X               X               X               X               X
netpol-5383-x/b X               X               X               X               X               X               X               X               X
netpol-5383-x/c X               X               X               X               X               X               X               X               X
netpol-5383-y/a .               .               .               .               .               .               .               .               .
netpol-5383-y/b .               .               .               .               .               .               .               .               .
netpol-5383-y/c .               .               .               .               .               .               .               .               .
netpol-5383-z/a .               .               .               .               .               .               .               .               .
netpol-5383-z/b .               .               .               .               .               .               .               .               .
netpol-5383-z/c .               .               .               .               .               .               .               .               .



Nov 15 10:29:28.203: INFO: observed:

-               netpol-5383-x/a netpol-5383-x/b netpol-5383-x/c netpol-5383-y/a netpol-5383-y/b netpol-5383-y/c netpol-5383-z/a netpol-5383-z/b netpol-5383-z/c
netpol-5383-x/a X               X               X               X               X               X               X               X               X
netpol-5383-x/b X               X               X               X               X               X               X               X               X
netpol-5383-x/c X               X               X               X               X               X               X               X               X
netpol-5383-y/a .               .               .               .               .               .               .               .               .
netpol-5383-y/b .               .               .               .               .               .               .               .               .
netpol-5383-y/c .               .               .               .               .               .               .               .               .
netpol-5383-z/a .               .               .               .               .               .               .               .               .
netpol-5383-z/b .               .               .               .               .               .               .               .               .
netpol-5383-z/c .               .               .               .               .               .               .               .               .



Nov 15 10:29:28.203: INFO: comparison:

-               netpol-5383-x/a netpol-5383-x/b netpol-5383-x/c netpol-5383-y/a netpol-5383-y/b netpol-5383-y/c netpol-5383-z/a netpol-5383-z/b netpol-5383-z/c
netpol-5383-x/a .               .               .               .               .               .               .               .               .
netpol-5383-x/b .               .               .               .               .               .               .               .               .
netpol-5383-x/c .               .               .               .               .               .               .               .               .
netpol-5383-y/a .               .               .               .               .               .               .               .               .
netpol-5383-y/b .               .               .               .               .               .               .               .               .
netpol-5383-y/c .               .               .               .               .               .               .               .               .
netpol-5383-z/a .               .               .               .               .               .               .               .               .
netpol-5383-z/b .               .               .               .               .               .               .               .               .
netpol-5383-z/c .               .               .               .               .               .               .               .               .



Nov 15 10:29:28.203: INFO: VALIDATION SUCCESSFUL
[AfterEach] NetworkPolicy between server and client
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/netpol/network_policy.go:138
Nov 15 10:29:28.203: INFO: DnsDomain cluster.local
[AfterEach] [sig-network] Netpol [LinuxOnly]
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:175
Nov 15 10:29:28.235: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "netpol-5383" for this suite.

• [SLOW TEST:3089.460 seconds]
[sig-network] Netpol [LinuxOnly]
/Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/framework.go:23
  NetworkPolicy between server and client
  /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/netpol/network_policy.go:114
    should support egress to the outside world while blocking internal pods [Feature:NetworkPolicy]
    /Users/jayunit100/SOURCE/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/netpol/network_policy.go:853
------------------------------
{"msg":"PASSED [sig-network] Netpol [LinuxOnly] NetworkPolicy between server and client should support egress to the outside world while blocking internal pods [Feature:NetworkPolicy]","total":1,"completed":1,"skipped":3725,"failed":0}
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSNov 15 10:29:28.267: INFO: Running AfterSuite actions on all nodes
Nov 15 10:29:28.267: INFO: Running AfterSuite actions on node 1
{"msg":"Test Suite completed","total":1,"completed":1,"skipped":5335,"failed":0}

```

### CODE REVIEW

**Comment by jayunit100** [comment]
I suspect @aojea will get a tickle out of this one .

**Comment by jayunit100** [comment]
This needs to wait until the new netpol validation stuff merges in

**Comment by jayunit100** [comment]
xref #91592

**Comment by jayunit100** [comment]
@vpickard , this is a new test case, following up on 9152, which exemplifies how 0.0.0.0/0 policies work when Denying all in-cluster traffic.... lmk what you think (just look at the last commit)

**Comment by jayunit100** [comment]
per history on this PR, i did it in JSON bc its a specific use case folks are confused about ... am not against doing it in golang, but figure the Json was better here , unless theres strong opinions about golang uniformity.

File: test/e2e/network/netpol/network_policy.go [hunk 1]
```diff
@@ -988,7 +989,80 @@ var _ = SIGDescribeCopy("Netpol [Feature:SCTPConnectivity][LinuxOnly][Disruptive
...
+			framework.ExpectNoError(err, "unmarshal network policy")
+
+			CreatePolicy(k8s, np, nsX)
+
+			reachability := NewReachability(model.AllPods(), true)
+			reachability.ExpectAllEgress(NewPodString(nsX, "a"), false)
+			reachability.ExpectAllEgress(NewPodString(nsX, "b"), false)
+			reachability.ExpectAllEgress(NewPodString(nsX, "c"), false)
+			fmt.Println("sleeeeep")
+			time.Sleep(50 * time.Minute)
```
Comment (mattfenwick) [inline test/e2e/network/netpol/network_policy.go:1060]: Is this sleep necessarly?  Definitely needs to be shortened

File: test/e2e/network/netpol/network_policy.go [hunk 2]
```diff
@@ -988,7 +989,80 @@ var _ = SIGDescribeCopy("Netpol [Feature:SCTPConnectivity][LinuxOnly][Disruptive
...
+		 }
+		 `, allowAll, podServerCIDR, allowAll)
+
+			err = json.Unmarshal([]byte(policy), np)
+			framework.ExpectNoError(err, "unmarshal network policy")
+
+			CreatePolicy(k8s, np, nsX)
+
+			reachability := NewReachability(model.AllPods(), true)
+			reachability.ExpectAllEgress(NewPodString(nsX, "a"), false)
```
Comment (mattfenwick) [inline test/e2e/network/netpol/network_policy.go:1056]: I think these 3 lines can be simplified to something like

```
ExpectPeer(&Peer{Namespace: nsX}, &Peer{}, false)
```

File: test/e2e/network/netpol/network_policy.go [hunk 3]
```diff
@@ -988,7 +989,80 @@ var _ = SIGDescribeCopy("Netpol [Feature:SCTPConnectivity][LinuxOnly][Disruptive
...
+
+		// This policy came up in https://github.com/kubernetes/kubernetes/issues/96341.   Its a good test case b/c it confirms that 0.0.0.0/0 works properly.
+		ginkgo.It("should support egress to the outside world while blocking internal pods [Feature:NetworkPolicy]", func() {
+			np := &networkingv1.NetworkPolicy{}
+
+			nsX, _, _, model, k8s := getK8SModel(f)
+			// grab a pods IP address so we can block everything in its subnet...
+			podList, err := f.ClientSet.CoreV1().Pods(nsX).List(context.TODO(), metav1.ListOptions{LabelSelector: "pod=a"})
+			framework.ExpectNoError(err, "Failing to find pod x/a")
+			podA := podList.Items[0]
```
Comment (mattfenwick) [inline test/e2e/network/netpol/network_policy.go:1001]: Rename to `podXA`

File: test/e2e/network/netpol/network_policy.go [hunk 4]
```diff
@@ -988,7 +989,80 @@ var _ = SIGDescribeCopy("Netpol [Feature:SCTPConnectivity][LinuxOnly][Disruptive
...
+
+			// Exclude podServer's IP with an Except clause
+			hostMask := 24
+			allowAll := "0.0.0.0/0"
+			if utilnet.IsIPv6String(podA.Status.PodIP) {
+				hostMask = hostMask * 4
```
Comment (mattfenwick) [inline test/e2e/network/netpol/network_policy.go:1007]: would `hostMask = 96` be clearer ?

File: test/e2e/network/netpol/network_policy.go [hunk 5]
```diff
@@ -988,7 +989,80 @@ var _ = SIGDescribeCopy("Netpol [Feature:SCTPConnectivity][LinuxOnly][Disruptive
 
 			ValidateOrFail(k8s, model, &TestCase{FromPort: 81, ToPort: 80, Protocol: v1.ProtocolSCTP, Reachability: reachability})
 		})
+
+		// This policy came up in https://github.com/kubernetes/kubernetes/issues/96341.   Its a good test case b/c it confirms that 0.0.0.0/0 works properly.
```
Comment (mattfenwick) [inline test/e2e/network/netpol/network_policy.go:993]: I think this test case is subtle and needs a better explanation.  Looking at the policy, I don't understand what this is testing, what the expected result is, and what assumptions it's making.

Can you add a comment explaining why the policy does what it does, and maybe add the truthtables from the PR description in the comment as well?

**Review by mattfenwick — COMMENTED** [review]
Interesting test case!  Needs a bit more explanation for people to grok it!

**Comment by knabben** [comment]
> per history on this PR, i did it in JSON bc its a specific use case folks are confused about ... am not against doing it in golang, but figure the Json was better here , unless theres strong opinions about golang uniformity.

Lets use the golang for uniformity and reusability. 
There's an example on this comment - https://github.com/kubernetes/kubernetes/issues/96341#issuecomment-724271981

**Pull request closed without merging**